rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,572 @@
+      SE_Event event, TF_Status* status);
+
+  // Wait for the specified event at the end of the specified stream.
+  void (*wait_for_event)(
+      SE_Device* executor, SE_Stream stream,
+      SE_Event event, TF_Status* status);
+
+  /*** TIMER CALLBACKS ***/
+  // Creates TF_Timer. Allocates timer resources on the underlying platform and initializes its
+  // internals.
+  void (*create_timer)(SE_Device* executor, SE_Timer* timer, SE_TimerFns** timer_fns, TF_Status* status);
+
+  // Destroy timer and deallocates timer resources on the underlying platform.
+  void (*destroy_timer)(SE_Device* executor, SE_Timer timer, SE_TimerFns* timer_fns);
+
+  // Records a start event for an interval timer.
+  TF_BOOL (*start_timer)(
+      SE_Device* executor, SE_Stream stream, SE_Timer timer);
+
+
+  // Records a stop event for an interval timer.
+  TF_BOOL (*stop_timer)(
+      SE_Device* executor, SE_Stream stream, SE_Timer timer);
+
+  /*** MEMCPY CALLBACKS ***/
+  // Enqueues a memcpy operation onto stream, with a host destination location
+  // host_dst and a device memory source, with target size size.
+  TF_BOOL (*memcpy_to_host)(
+      SE_Device* executor, SE_Stream stream,
+      void* host_dst,
+      const SE_DeviceMemoryBase* device_src,
+      uint64_t size);
+
+  // Enqueues a memcpy operation onto stream, with a device destination location
+  // and a host memory source, with target size size
+  TF_BOOL (*memcpy_from_host)(
+      SE_Device* executor, SE_Stream stream,
+      SE_DeviceMemoryBase* device_dst,
+      const void* host_src, uint64_t size);
+
+  // Causes the host code to synchronously wait for operations enqueued onto
+  // stream to complete. Effectively a join on the asynchronous device
+  // operations enqueued on the stream before this program point.
+  void (*block_host_until_done)(
+      SE_Device* executor, SE_Stream stream,
+      TF_Status* status);
+
+  // Synchronizes all activity occurring in the StreamExecutor's context (most
+  // likely a whole device).
+  TF_BOOL (*synchronize_all_activity)(SE_Device* executor);
+
+  // Obtains metadata about the underlying device.
+  void (*fill_device_description)(SE_Device* executor,
+                               SE_DeviceDescription* description,
+                               TF_Status* status);
+
+  // Enqueues on a stream a user-specified function to be run on the host.
+  TF_BOOL (*host_callback)(SE_Device* executor, SE_Stream* stream,
+                     TF_StatusCallbackFn callback_fn, void* ctx);
+} SE_StreamExecutor;
+
+#define SE_STREAMEXECUTOR_STRUCT_SIZE TF_OFFSET_OF_END(SE_StreamExecutor, host_callback)
+
+typedef struct SE_PlatformParams {
+  size_t struct_size;
+  
+  // Free form data set by plugin. It will be pre-filled
+  // in SE_Device passed to create_device.
+  void* data;
+  SE_PlatformId* id;
+  
+  // Platform name
+  const char* name;
+  size_t name_len;
+  
+  // Device type name. Right now only GPU is supported.
```

Comment:
I understand this Device type name is the front-end name(tf.device), right? why only "GPU" is supported?
and do you prefer device_type attribute of class Device( member: const DeviceAttributes device_attributes_)  such as (DEVICE_GPU) also be registered from plugin through this API? 
```
BaseGPUDevice::BaseGPUDevice(const SessionOptions& options, const string& name,
                             Bytes memory_limit, const DeviceLocality& locality,
                             TfGpuId tf_gpu_id,
                             const string& physical_device_desc,
                             Allocator* gpu_allocator, Allocator* cpu_allocator,
                             bool sync_every_op)
    : LocalDevice(options, Device::BuildDeviceAttributes(name, DEVICE_GPU,
                                                         memory_limit, locality,
                                                         physical_device_desc)),
```

This device_type will be used to register the kernel and do branch in the common code, if the device type is defined in tensorflow, then multiple-plugins will be an issue since only one kernel can be registered successfully, such as 
```
// Registration of the GPU implementations.
REGISTER_KERNEL_BUILDER(
    Name("Conv2D").Device(DEVICE_GPU).TypeConstraint<Eigen::half>("T"),
    Conv2DOp<GPUDevice, Eigen::half>);
```
if this type can be registered from plugin, we may also need to support the branch code in TF,
```
if (device_type != DEVICE_GPU && device_type != DEVICE_SYCL && !IsPluggedType(device_type)) {
     // On non-GPU and non-SYCL devices, HOST_MEMORY and DEVICE_MEMORY are always
     // compatible.
     return Status::OK();
   }  
```

Reply:
We can allow any type name. I was just thinking that PluggableDevice is going to be GPU-specific. If it is not, then we can allow passing arbitrary name.